### PR TITLE
Implement some of api:validation:vertex_state

### PR DESCRIPTION
**Pull request by Kangz**
PTAL, this implements some of the test plan, but not tests that are waiting for group decision or tests that use formats.

Should I add something like the texture format info, but for vertex formats to the CTS? This will help make parametrizations easier to do, but is more upfront complexity.

-----

- [X] New helpers, if any, are documented in `helper_index.md`.
- [X] Incomplete tests, if any, are marked with TODO or `.unimplemented()`.

**[Review requirements](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) are satisfied for:**

- [ ] WebGPU readability
- [ ] TypeScript readability

### CODE REVIEW

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 1]
```diff
@@ -55,22 +38,22 @@ Test that an attribute must be aligned to the component size:
...
+import { params, pbool, poptions } from '../../../common/framework/params_builder.js';
 import { makeTestGroup } from '../../../common/framework/test_group.js';
 
 import { ValidationTest } from './validation_test.js';
 
 const MAX_VERTEX_ATTRIBUTES: number = 16;
 const MAX_VERTEX_BUFFER_END: number = 2048;
 const MAX_VERTEX_BUFFER_ARRAY_STRIDE: number = 2048;
-const MAX_VERTEX_BUFFERS: number = 16;
+const MAX_VERTEX_BUFFERS: number = 8;
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:49]: Consider moving all of these constants into capability_info (and renaming to kMaxVertexBuffers etc). BTW the type annotation is unnecessary.

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 2]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+      vertexBuffers.push({ arrayStride: 0, attributes });
+    }
+
+    const success = attribCount <= MAX_VERTEX_ATTRIBUTES;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('max_vertex_buffer_array_stride_limit')
+  .desc(
+    `Test that the vertex buffer arrayStride must be less than <maxVertexBufferArrayStride>
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:193]:
```suggestion
    `Test that the vertex buffer arrayStride must be at most <maxVertexBufferArrayStride>.
```

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 3]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+    const vertexBuffers: GPUVertexBufferLayoutDescriptor[] = [];
+    vertexBuffers[vertexBufferIndex] = { arrayStride, attributes: [] };
+
+    const success = arrayStride <= MAX_VERTEX_BUFFER_ARRAY_STRIDE;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('vertex_buffer_array_stride_limit_alignment')
+  .desc(
+    `Test that the vertex buffer arrayStride must be aligned to 4
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:223]:
```suggestion
    `Test that the vertex buffer arrayStride must be a multiple of 4 (including 0).
```

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 4]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+   - Test for various vertex buffer indices
+   - Test for array strides 0, 1, 2, 4, limit - 4, limit -2, limit`
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:225]:
```suggestion
   - Test for array strides 0, 1, 2, 4, limit - 4, limit - 2, limit`
```

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 5]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+    const vertexBuffers: GPUVertexBufferLayoutDescriptor[] = [];
+    vertexBuffers[vertexBufferIndex] = { arrayStride, attributes: [] };
+
+    const success = arrayStride % 4 === 0;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('vertex_attribute_shaderLocation_limit')
+  .desc(
+    `Test that the vertex buffer arrayStride must be aligned to 4
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:254]:
```suggestion
    `Test shaderLocation must be less than maxVertexAttributes.
```

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 6]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+   - Test for various vertex buffer indices
+   - Test for various amounts of attributes in that vertex buffer
+   - Test for shaderLocation 0, 1, limit -1, limit (limit is maxVertexAttributes)`
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:257]:
```suggestion
   - Test for shaderLocation 0, 1, limit - 1, limit`
```

File: src/webgpu/api/validation/vertex_state.spec.ts [hunk 7]
```diff
@@ -105,87 +88,220 @@ class F extends ValidationTest {
...
+  )
+  .subcases(() =>
+    params()
+      .combine(poptions('vertexBufferIndex', [0, 1, MAX_VERTEX_BUFFERS - 1]))
+      .combine(poptions('attributeCount', [0, 1, MAX_VERTEX_ATTRIBUTES - 1]))
```
Comment (kainino0x) [inline src/webgpu/api/validation/vertex_state.spec.ts:262]: nit: "extraAttributes"? since this doesn't include the test attribute

**Review by kainino0x — APPROVED** [review]

**Comment by kainino0x** [comment]
> Should I add something like the texture format info, but for vertex formats to the CTS? This will help make parametrizations easier to do, but is more upfront complexity.

I think you should. They probably won't be used in a ton of tests, but probably worth it anyway. Tables are easier to maintain and I don't think they'll add that much complexity.

File: src/webgpu/api/validation/vertex_state.spec.ts (hunk 1 above)
Comment (Kangz) [reply src/webgpu/api/validation/vertex_state.spec.ts:49]: I'll do that as a follow-up.

**Comment by Kangz** [comment]
Thanks!

**Pull request merged**